### PR TITLE
Add release guide

### PR DESCRIPTION
**Pull request by lewtun**
This PR adds a small release guide based on what we use internally at Hugging Face :)

### CODE REVIEW

File: RELEASE.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
+# A guide to making a release
+
+This guide collects the steps we do in SetFit to make a release on PyPI. They result from hard-learned lessons (mostly on Transformers) and while following this guide is completely optional, it’s strongly recommended to do so. 🙂
+
+### Preparation
+
+To be able to make a release for a given project, you’ll need an account on [PyPI](https://pypi.org/) and on [Test PyPI](https://test.pypi.org/). If you are making a release for an existing project, your username will need to be added to that project by one of the current maintainers on PypI. Note that we strongly recommend enabling two-factor authentication on PyPI.
```
Comment (tomaarsen) [inline RELEASE.md:7]:
```suggestion
To be able to make a release for a given project, you’ll need an account on [PyPI](https://pypi.org/) and on [Test PyPI](https://test.pypi.org/). If you are making a release for an existing project, your username will need to be added to that project by one of the current maintainers on PyPI. Note that we strongly recommend enabling two-factor authentication on PyPI.
```

Comment (tomaarsen) [reply RELEASE.md:7]: PyPI forces me to use 2FA due to NLTK 😅

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+git push -u origin v{major}.{minor}-release
+```
+
+**Make sure you are at the right commit on the main branch before executing this and do not forget to pull the main branch if you want to branch from the last commit!**
+
+### Step 2: Prepare the release notes
+
+We have [a space](https://huggingface.co/spaces/lysandre/github-release) to help you with that. You will need a [GitHub token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) with the following scopes:
+
+![](assets/scopes.png)
```
Comment (tomaarsen) [inline RELEASE.md:32]: I tend to drag and drop or paste files into GitHub to do this. This results in:

`image`

which produces

image

Then I don't have to use an assets folder. Is there any reason for using the assets folder? Just so you know it'll always be there?

File: RELEASE.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+Save the token somewhere safe, as GitHub does not let you see it again after it’s created. Next, adjust the settings of the Space to get the whole list of commits done since last release, along with the top contributors outside of the Hugging Face team.
+
+You can then put your release notes in a Draft Release on GitHub, in [https://github.com/huggingface/setfit/releases](https://github.com/huggingface/setfit/releases) (replace project_name by the package you are working on) and write a small paragraph highlighting each of the new features this release is adding.
```
Comment (tomaarsen) [inline RELEASE.md:36]:
```suggestion
You can then put your release notes in a Draft Release on GitHub, in [https://github.com/huggingface/setfit/releases](https://github.com/huggingface/setfit/releases) and write a small paragraph highlighting each of the new features this release is adding.
```
Was this part of the template? Or should it be replaced in the Draft release?

File: RELEASE.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+### Step 3: Adjust the version of your package
+
+You should have the current version specified in the `[setup.py](http://setup.py)` and then the main `__init__.py` files. This version should be a dev version before you release, change it to the name of the version you are releasing by running:
```
Comment (tomaarsen) [inline RELEASE.md:40]:
```suggestion
You should have the current version specified in the `[setup.py](setup.py)` and then the main `__init__.py` files. This version should be a dev version before you release, change it to the name of the version you are releasing by running:
```

File: RELEASE.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+Here is how you can create and push your tag:
+
+```bash
+git tag v<VERSION> -m 'Adds tag v<VERSION> for PyPI'
+git push --tags origin v{major}.{minor}-release
+```
+
+### Step 6: Create the wheels for your release
+
+This is the stuff you will upload on PyPI and that everyone will then download each time they pip install your package.
```
Comment (tomaarsen) [inline RELEASE.md:72]:
```suggestion
This is what you'll upload on PyPI and what everyone will download each time they `pip install` your package.
```
This suggestion's a bit nitpicky.

File: RELEASE.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+To upload it:
+
+```bash
+make pypi_test_upload
+```
+
+You will be prompted for your username and password.
+
+Then in a fresh environment containing all dependencies you need, try to install your new package from the pypi test server
```
Comment (tomaarsen) [inline RELEASE.md:102]:
```suggestion
Then in a fresh environment containing all dependencies you need, try to install your new package from the PyPI test server.
```

File: RELEASE.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+```bash
+make pypi_upload
+```
+
+You will be prompted for your username and password.
+
+### Step 9: Publish your release notes
+
+Go back to the draft you did at step 2 ([https://github.com/huggingface/setfit/releases](https://github.com/huggingface/setfit/releases) with project_name replaced by the package you are working on) and add the tag you created at step 4 to your release before publishing your release notes
```
Comment (tomaarsen) [inline RELEASE.md:128]:
```suggestion
Go back to the draft you did at step 2 ([https://github.com/huggingface/setfit/releases](https://github.com/huggingface/setfit/releases)) and add the tag you created at step 4 to your release before publishing your release notes.
```

File: RELEASE.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+### Step 10: Bump the dev version on the main branch
+
+You’re almost done! Just go back to the `main` branch and change the dev version in the `setup.py` and the main `__init__.py` to the new version you’re developing, for instance `4.13.0.dev0` if just released `4.12.0`
```
Comment (tomaarsen) [inline RELEASE.md:132]:
```suggestion
You’re almost done! Just go back to the `main` branch and change the dev version in the `setup.py` and the main `__init__.py` to the new version you’re developing, for instance `4.13.0.dev0` if just released `4.12.0`.
```

File: RELEASE.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+Then cherry-pick the commit fixing the bug for which you are making the patch release. You will need to find its commit hash.
+
+```bash
+git cherry-pick {commit_hash}
+git push -u origin v{major}.{minor}-release
+```
+
+### Step 2: Adjust the version of your package
+
+You should have the current version specified in the `[setup.py](http://setup.py)` and then the main `__init__.py` files. This should be a bump in the micro version (so go from `4.12.0` to `4.12.1` for instance). Tthere is a utility to do this step with `make pre-patch`.
```
Comment (tomaarsen) [inline RELEASE.md:157]:
```suggestion
You should have the current version specified in the `[setup.py](setup.py)` and then the main `__init__.py` files. This should be a bump in the micro version (so go from `4.12.0` to `4.12.1` for instance). There is a utility to do this step with `make pre-patch`.
```

File: RELEASE.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+Go back to the draft you did at step 2 ([https://github.com/huggingface/setfit/releases](https://github.com/huggingface/setfit/releases) with project_name replaced by the package you are working on) and add the tag you created at step 4 to your release before publishing your release notes
+
+### Step 10: Bump the dev version on the main branch
+
+You’re almost done! Just go back to the `main` branch and change the dev version in the `setup.py` and the main `__init__.py` to the new version you’re developing, for instance `4.13.0.dev0` if just released `4.12.0`
+
+### Patch releases
+
+Follow this guide when you are releasing a patch to fix a bug introduced in a release. For minor or major releases go [here](https://www.notion.so/Make-a-release-81fbe41ad0e24657a3126dfc7fd1c4b6).
```
Comment (tomaarsen) [inline RELEASE.md:136]:
```suggestion
Follow this guide when you are releasing a patch to fix a bug introduced in a release. For minor or major releases go [here](#minor-or-major-releases).
```
I think this should work, but I'm not sure.

File: RELEASE.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+Here is how you can create and push your tag:
+
+```bash
+git tag v<VERSION> -m 'Adds tag v<VERSION> for pypi'
+git push --tags origin v{major}.{minor}-release
+```
+
+### Step 5: Create the wheels for your release
+
+This is the stuff you will upload on PyPI and that everyone will then download each time they pip install your package.
```
Comment (tomaarsen) [inline RELEASE.md:185]:
```suggestion
This is what you'll upload on PyPI and what everyone will download each time they `pip install` your package.
```

File: RELEASE.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+To upload it:
+
+```bash
+make pypi_test_upload
+```
+
+You will be prompted for your username and password.
+
+Then in a fresh environment containing all dependencies you need, try to install your new package from the pypi test server
```
Comment (tomaarsen) [inline RELEASE.md:215]:
```suggestion
Then in a fresh environment containing all dependencies you need, try to install your new package from the PyPI test server.
```

File: RELEASE.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+This cannot be undone if you messed up, so make sure you have run Step 7!
+
+Once you’re fully ready, upload your package on PyPI:
+
+```bash
+make pypi_upload
+```
+
+You will be prompted for your username and password.
```
Comment (tomaarsen) [inline RELEASE.md:239]:
```suggestion
You will be prompted for your username and password, unless you're using the recommended [PyPI API token](https://pypi.org/help/#apitoken).
```

File: RELEASE.md [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+### Step 8: Publish your release notes
+
+Go to your GitHub release pages ([https://github.com/huggingface/project-name/releases](https://github.com/huggingface/project-name/releases) with project_name replaced by the package you are working on) and write a quick explanation of what the patch fixes. Include the tag you added at step 4 and you are good to publish!
```
Comment (tomaarsen) [inline RELEASE.md:243]:
```suggestion
Go to your GitHub release pages ([https://github.com/huggingface/setfit/releases](https://github.com/huggingface/setfit/releases)) and write a quick explanation of what the patch fixes. Include the tag you added at step 4 and you are good to publish!
```

**Review by tomaarsen — CHANGES_REQUESTED** [review]
I've got a few small nitpicks.
In hindsight, it's more than a few, that wasn't really the intention haha.

File: RELEASE.md (hunk 2 above)
Comment (lewtun) [reply RELEASE.md:32]: Oh this is a cool trick! Yeah, I usually have an `assets` folder for images (especially if they will change), but in this case a screenshot hosted on GitHub is much better

File: RELEASE.md (hunk 3 above)
Comment (lewtun) [reply RELEASE.md:36]: Ah nice catch! Yes, this was part of a more generic guide we have internally :)

File: RELEASE.md [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+This cannot be undone if you messed up, so make sure you have run Step 7!
+
+Once you’re fully ready, upload your package on PyPI:
+
+```bash
+make pypi_upload
+```
+
+You will be prompted for your username and password.
```
[the remark this answers is not in the document]
Comment (lewtun) [reply RELEASE.md:124]: Nice, I didn't know about this - definitely a nicer approach!

File: RELEASE.md (hunk 10 above)
Comment (lewtun) [reply RELEASE.md:136]: Nice catch!

File: RELEASE.md (hunk 2 above)
Comment (lewtun) [reply RELEASE.md]: Done in https://github.com/huggingface/setfit/pull/357/commits/6a67da0723f5578203a7ee61097eeff06c4d290b

**Comment by lewtun** [comment]
Thanks for the careful review @tomaarsen ! I've included all your suggestions and also switched the screenshot to use your image URL :)

**Comment by tomaarsen** [comment]
Looks great! Thanks for the detailed guide, it'll be helpful for sure.

**Pull request merged**